eventuals/tcp.h

```diff
@@ -0,0 +1,583 @@
+                  asio::ip::tcp::endpoint endpoint;
+
+                  switch (context.socket->protocol_) {
+                    case Protocol::IPV4:
+                      endpoint = asio::ip::tcp::endpoint(
+                          asio::ip::make_address_v4(context.ip, ec),
+                          context.port);
+                      break;
+                    case Protocol::IPV6:
+                      endpoint = asio::ip::tcp::endpoint(
+                          asio::ip::make_address_v6(context.ip, ec),
+                          context.port);
+                      break;
+                    default:
+                      k.Fail("Unknown protocol");
+                      return;
+                  }
+
+                  if (ec) {
+                    k.Fail(ec.message());
+                    return;
+                  }
+
+                  context.socket->underlying_handle().bind(endpoint, ec);
+
+                  if (!ec) {
+                    k.Start();
+                  } else {
+                    k.Fail(ec.message());
+                  }
+                }
+              });
+        });
+  }
+
+  auto Connect(std::string&& ip, uint16_t port) {
+    return RescheduleAfter(_Connect::Composable{*this, ip, port});
+  }
+
+  auto Close() {
+    return Eventual<void>()
+        .interruptible()
+        .context(this)
+        .start([](Socket* socket, auto& k, Interrupt::Handler& handler) {
+          asio::post(
+              socket->io_context(),
+              [socket,
+               &k,
+               &handler]() {
+                if (handler.interrupt().Triggered()) {
+                  k.Stop();
+                } else {
+                  asio::error_code ec;
+
+                  socket->underlying_handle().close(ec);
+
+                  if (!ec) {
+                    k.Start();
+                  } else {
+                    k.Fail(ec.message());
+                  }
+                }
+              });
+        });
+  }
+
+  // Maybe thread-unsafe.
+  // If there is any other operation on acceptor
+  // on a different thread, return value might be incorrect.
+  bool IsOpen() {
+    return underlying_handle().is_open();
+  }
+
+  // Maybe thread-unsafe.
+  // If there is any other operation on acceptor
+  // on a different thread, return value might be incorrect.
+  uint16_t BindedPort() {
+    return underlying_handle().local_endpoint().port();
+  }
+
+  // Maybe thread-unsafe.
+  // If there is any other operation on acceptor
+  // on a different thread, return value might be incorrect.
+  std::string BindedIp() {
+    return underlying_handle().local_endpoint().address().to_string();
+  }
+
+ private:
+  EventLoop& loop_;
+  asio::ip::tcp::socket socket_;
+  Protocol protocol_ = Protocol::NOT_SET;
+
+  asio::ip::tcp::socket& underlying_handle() {
+    return socket_;
+  }
+
+  asio::io_context& io_context() {
+    return loop_.io_context();
+  }
+
+  struct _Connect {
```

Comment:
It's definitely not obvious why you can just use `Eventual` here as well? It'll remove having to implement `Fail()` and `Stop()` which are just copied code.